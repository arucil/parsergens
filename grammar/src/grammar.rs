use std::ops::Range;
use std::borrow::Borrow;
use super::lexer::{Lexer, TokenId};
use crate::{Set, BiMap, Map};

pub use lower::*;

pub mod lower;

#[derive(Debug)]
pub struct Grammar {
  /// productions of the same nonterminals are consecutive.
  pub rules: Vec<Rule>,
  /// starting non-terminals
  pub start_nts: Set<NonterminalId>,
  /// non-terminals
  pub nts: BiMap<NonterminalId, String>,
  pub nt_prods: Map<NonterminalId, Range<usize>>,
  pub lexer: Lexer,
}

#[derive(Debug)]
pub struct Rule {
  pub nt: NonterminalId,
  pub items: Vec<Item>,
}

#[derive(Debug)]
pub enum Item {
  Nonterminal(NonterminalId),
  Token(TokenId),
  Optional(Vec<Item>),
  Many(Vec<Item>),
  Many1(Vec<Item>),
}

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
pub struct NonterminalId(u32);

impl NonterminalId {
  pub fn id(&self) -> u32 {
    self.0
  }
}

impl Borrow<u32> for NonterminalId {
  fn borrow(&self) -> &u32 {
    &self.0
  }
}

#[derive(Default)]
pub struct NonterminalIdGen(u32);

impl NonterminalIdGen {
  pub fn gen(&mut self) -> NonterminalId {
    let i = self.0;
    self.0 += 1;
    NonterminalId(i)
  }

  pub fn from(start: u32) -> Self {
    Self(start + 1)
  }
}

impl Grammar {
  pub fn lower(self) -> LoweredGrammar {
    lower::lower(self)
  }
}