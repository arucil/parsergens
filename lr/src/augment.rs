use grammar::NonterminalIdGen;
use grammar::{
  LoweredGrammar, Production, ProductionKind, Symbol,
  LoweredNonterminal, NonterminalKind,
};

/// Add `S' -> S $` to grammar, where `$` is a new token representing EOF.
pub fn augment(grammar: LoweredGrammar) -> LoweredGrammar {
  let max_nt_id = grammar.nts
    .keys()
    .map(|x| x.id())
    .max()
    .unwrap();
  let mut nt_id_gen = NonterminalIdGen::from(max_nt_id);

  let mut prods = grammar.prods;
  let mut nts = grammar.nts;

  let start_nts = grammar.start_nts.into_iter().map(|nt| {
    let new_start_nt = nt_id_gen.gen();
    let start = prods.len();

    prods.push(Production {
      nt: new_start_nt,
      kind: ProductionKind::Ordinary,
      symbols: vec![
        Symbol::Nonterminal(nt),
        //Symbol::Token(eof_token),
      ],
      prec: None,
      action: None,
    });

    let new_nt_name = format!("_{}", nts[&nt].name);

    nts.insert(new_start_nt, LoweredNonterminal {
      name: new_nt_name,
      range: start..prods.len(),
      ty: None,
      kind: NonterminalKind::User,
    });

    new_start_nt
  }).collect();

  let grammar = LoweredGrammar {
    prods,
    start_nts,
    nts,
    ..grammar
  };

  grammar
}
