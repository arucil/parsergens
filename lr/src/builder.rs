use grammar::{LoweredGrammar, Symbol, Map};
use indexmap::IndexMap;
use fnv::FnvBuildHasher;
use std::fmt::{self, Write};
use std::hash::Hash;
use crate::token_set::TokenSet;

pub use self::states::gen_states;
pub use self::tables::gen_tables;
pub use self::compress::{compress_tables, CompressedTables};

mod states;
mod tables;
mod compress;

pub struct Builder<'a, T: LrComputation> {
  pub grammar: &'a LoweredGrammar,
  pub states: StateStore<T::StateKey>,
  /// eof is the token with the greatest id
  pub eof: usize,
  /// max number of RHS symbols in productions, plus one.
  pub max_nsym_p1: usize,
}

pub type StateStore<Key> = IndexMap<Key, State, FnvBuildHasher>;

/// sorted.
pub type KernelItemSet = Vec<Item>;

pub type LookaheadSet = TokenSet;

#[derive(Debug)]
pub struct State {
  /// Starts with sorted `kernel_len` kernel items, followed by other non-kernel items.
  pub items: Vec<Item>,
  /// size of kernel item set
  pub kernel_len: usize,
  /// symbol -> index of destination state
  pub transitions: Map<Symbol, u32>,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Item {
  /// production and dot
  pub key: u32,
  pub lookaheads: LookaheadSet,
}

pub trait LrComputation {
  type StateKey: Eq + Hash;

  /// Returns the index of the state and whether the state has changed.
  fn store_state(
    states: &mut StateStore<Self::StateKey>,
    kernel_item_set: KernelItemSet,
  ) -> (u32, bool);
}

fn encode_item(
  max_nsym_p1: usize,
  prod_ix: usize,
  dot: usize,
) -> u32 {
  (prod_ix * max_nsym_p1 + dot) as u32
}

fn decode_item(
  max_nsym_p1: usize,
  key: u32,
) -> (usize, usize) {
  let prod = key as usize / max_nsym_p1;
  let dot = key as usize % max_nsym_p1;
  (prod, dot)
}

impl State {
  pub fn new(items: KernelItemSet) -> Self {
    State {
      kernel_len: items.len(),
      items,
      transitions: Map::default(),
    }
  }
}

impl<'a, T: LrComputation> Builder<'a, T> {
  pub fn new(
    grammar: &'a LoweredGrammar
  ) -> Self {
    Self {
      grammar,
      states: StateStore::default(),
      eof: grammar.tokens.len(),
      max_nsym_p1: grammar.prods.iter()
        .map(|prod| prod.symbols.len())
        .max()
        .unwrap()
        + 1,
    }
  }
}

#[cfg(test)]
use grammar::HashMap;

#[cfg(test)]
use crate::EntryPoint;

#[cfg(test)]
impl<'a, T: LrComputation> Builder<'a, T> {
  pub fn states(
    &self,
    entry_points: &HashMap<String, EntryPoint>
  ) -> String {
    let mut output = String::new();
    self.fmt_states(entry_points, &mut output).unwrap();
    output
  }

  fn fmt_states(
    &self,
    entry_points: &HashMap<String, EntryPoint>,
    fmt: &mut impl Write,
  ) -> fmt::Result {
    for i in 0..self.states.len() {
      self.fmt_state(entry_points, i, fmt)?;
    }

    Ok(())
  }

  fn fmt_state(
    &self,
    entry_points: &HashMap<String, EntryPoint>,
    state_ix: usize,
    fmt: &mut impl Write,
  ) -> fmt::Result {
    let state = &self.states[state_ix];
    write!(fmt, "State {}", state_ix)?;
    if entry_points.values().find(|x| x.start_state == state_ix as u32).is_some() {
      write!(fmt, " (start)")?;
    }
    writeln!(fmt)?;

    for item in &state.items {
      self.fmt_item(item, fmt)?;

      writeln!(fmt)?;
    }

    writeln!(fmt)
  }
}

impl<'a, T: LrComputation> Builder<'a, T> {
  fn fmt_item(
    &self,
    item: &Item,
    fmt: &mut impl Write,
  ) -> fmt::Result {
    //let item = self.item_store.items[item_ix];
    let (prod, dot) = decode_item(self.max_nsym_p1, item.key);
    let nt = self.grammar.prods[prod].nt;
    let symbols = &self.grammar.prods[prod].symbols;

    write!(fmt, "{} ->", self.grammar.nts[&nt].name)?;

    for (i, sym) in symbols.iter().enumerate() {
      if i == dot {
        write!(fmt, " .")?;
      }

      match sym {
        Symbol::Token(token) => {
          let name = self.grammar.tokens.get(token).map(|s|s.as_str()).unwrap_or("$");
          write!(fmt, " {}", name)?;
        }
        Symbol::Nonterminal(nt) => {
          let name = &self.grammar.nts[nt].name;
          write!(fmt, " {}", name)?;
        }
      }
    }

    if dot == symbols.len() {
      write!(fmt, " .")?;
    }

    write!(fmt, "      ")?;

    let mut slash = false;
    for lookahead in item.lookaheads.iter() {
      if slash {
        write!(fmt, " / ")?;
      }
      slash = true;

      write!(fmt, "{}",
        self.grammar.tokens.get(&lookahead).map(|s|s.as_str()).unwrap_or("$"))?;
    }

    Ok(())
  }
}