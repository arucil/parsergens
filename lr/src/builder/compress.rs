use grammar::{LoweredGrammar, HashMap};
use std::hash::Hash;
use num_traits::int::PrimInt;

#[derive(Debug, Default)]
pub struct CompressedTables {
  /// ACTION table + GOTO table
  pub parse_table: Vec<i32>,

  /// displacement values for rows of ACTION table.
  pub action_disp: Vec<isize>,

  /// displacement values for rows of GOTO table.
  pub goto_disp: Vec<isize>,

  /// default values for rows of ACTION table.
  pub action_default: Vec<i32>,

  /// default values for rows of GOTO table.
  pub goto_default: Vec<i32>,

  /// -1 in `check_table` represents an empty entry.
  pub check_table: Vec<i32>,
}

#[derive(Debug, Default)]
struct CompressState {
  parse_table: Vec<i32>,
  check_table: Vec<i32>,
  first_empty: usize,
}

pub fn compress_tables(
  grammar: &LoweredGrammar,
  action: Vec<Vec<i32>>,
  goto: Vec<Vec<u32>>,
) -> CompressedTables {
  let mut state = CompressState::default();
  let num_tokens = grammar.tokens.len() + 1; // includes EOF

  let (action_disp, action_default) = compress_action_table(
    &mut state,
    action,
  );

  let (goto_disp, goto_default) = compress_goto_table(
    &mut state,
    goto,
    num_tokens,
  );

  while state.check_table.last() == Some(&-1) {
    state.check_table.pop();
    state.parse_table.pop();
  }

  for x in &mut state.parse_table {
    if *x == super::tables::ERROR_ACTION {
      *x = 0;
    }
  }

  CompressedTables {
    parse_table: state.parse_table,
    check_table: state.check_table,
    action_disp,
    action_default,
    goto_disp,
    goto_default,
  }
}

/// Returns displacement table and default value table.
fn compress_action_table(
  state: &mut CompressState,
  action: Vec<Vec<i32>>,
) -> (Vec<isize>, Vec<i32>) {
  let mut action = action.into_iter()
    .enumerate()
    .map(|(i, row)| make_row(i, row, |x| x > 0 || x == super::tables::ERROR_ACTION))
    .collect::<Vec<_>>();

  let default_table = action.iter().map(|row| row.default).collect::<Vec<_>>();

  action.sort_by_key(|row| row.num_default);
  let action = action;

  let mut disp_map = HashMap::<&[i32], isize>::default();
  let mut disp_table = vec![0; action.len()];

  for row in &action {
    if let Some(&disp) = disp_map.get(&row.values[..]) {
      // merge rows with exactly the same actions.
      disp_table[row.index] = disp;
    } else {
      // check_base is token id
      let disp = write_row(state, row, 0, false, true);
      disp_map.insert(&row.values, disp);
      disp_table[row.index] = disp;
    }
  }

  (disp_table, default_table)
}

/// Returns displacement table and default value table.
fn compress_goto_table(
  state: &mut CompressState,
  goto: Vec<Vec<u32>>,
  num_tokens: usize,
) -> (Vec<isize>, Vec<i32>) {
  let mut goto = goto.into_iter()
    .enumerate()
    .map(|(i, row)| make_row(i, row, |_| false))
    .collect::<Vec<_>>();

  let default_table = goto.iter().map(|row| row.default as i32).collect::<Vec<_>>();

  goto.sort_by_key(|row| row.num_default);
  let goto = goto;

  let mut disp_table = vec![0; goto.len()];

  for row in &goto {
    // check_base is token NT + num_tokens
    let disp = write_row(state, row, num_tokens, true, false);
    disp_table[row.index] = disp;
  }

  (disp_table, default_table)
}

/// Returns the displacement.
fn write_row<T: PrimInt>(
  state: &mut CompressState,
  row: &Row<T>,
  check_base: usize,
  check_row: bool,
  merge_rows: bool,
) -> isize {
  let mut disp = state.first_empty as isize;

  for &v in &row.values {
    if v == T::zero() {
      disp -= 1;
    } else {
      break;
    }
  }

  if state.first_empty as isize - disp == row.values.len() as isize {
    return -(row.values.len() as isize);
  }

  if merge_rows {
    'find_disp: loop {
      let mut i = disp;
      for (j, &v) in row.values.iter().enumerate() {
        if i >= 0 && (i as usize) < state.check_table.len() &&
          state.check_table[i as usize] >= 0
        {
          let i = i as usize;
          let next = if v == T::zero() {
            // two rows that have unoverlapped non-zero values are not allowed
            // to be merged.
            state.check_table[i] == j as i32
          } else {
            state.check_table[i] != j as i32 || state.parse_table[i] != v.to_i32().unwrap()
          };
          if next {
            disp += 1;
            continue 'find_disp;
          }
        }
        i += 1;
      }
      break;
    }
  } else {
    'find_disp2: loop {
      let mut i = disp;
      for &v in &row.values {
        if !(v == T::zero() || i < 0 ||
          (i as usize) >= state.check_table.len() ||
          state.check_table[i as usize] < 0)
        {
          disp += 1;
          continue 'find_disp2;
        }
        i += 1;
      }
      break;
    }
  }

  let row_end = (disp + row.values.len() as isize) as usize;
  if row_end > state.check_table.len() {
    state.check_table.resize(row_end, -1);
    state.parse_table.resize(row_end, 0);
  }

  for (i, &v) in row.values.iter().enumerate() {
    if v != T::zero() {
      let j = (disp + i as isize) as usize;
      assert!(state.check_table[j] == -1);
      state.parse_table[j] = v.to_i32().unwrap();
      state.check_table[j] = if check_row {
        (check_base + row.index) as i32
      } else {
        (check_base + i) as i32
      };
    }
  }

  while state.first_empty < state.check_table.len() &&
    state.check_table[state.first_empty] == -1
  {
    state.first_empty += 1;
  }

  disp
}

fn make_row<T, F>(
  index: usize,
  mut values: Vec<T>,
  ignore_value: F,
) -> Row<T>
where
  T: Hash + PrimInt,
  F: Fn(T) -> bool,
{
  let mut count = HashMap::<T, usize>::default();
  let mut num_zeros = 0;

  for &v in &values {
    if v == T::zero() {
      num_zeros += 1;
    } else if !ignore_value(v) {
      *count.entry(v).or_default() += 1;
    }
  }

  if let Some((default, num_default)) = count.into_iter().max_by_key(|x| x.1) {
    // clear default values
    for v in &mut values {
      if *v == default {
        *v = T::zero();
      }
    }

    Row {
      index,
      default,
      num_default: num_default + num_zeros,
      values,
    }
  } else {
    Row {
      index,
      default: T::zero(),
      num_default: num_zeros,
      values,
    }
  }
}

#[derive(Debug, PartialEq, Eq)]
struct Row<T> {
  index: usize,
  default: T,
  num_default: usize,
  values: Vec<T>,
}

#[cfg(test)]
mod tests {
  use pretty_assertions::assert_eq;
  use super::*;

  #[test]
  fn make_row_all_zeros() {
    let vec = vec![0; 13];
    let row = make_row(1, vec.clone(), |_| false);
    assert_eq!(row, Row {
      index: 1,
      default: 0,
      num_default: 13,
      values: vec,
    });
  }

  #[test]
  fn make_row_most_frequent() {
    let vec1 = vec![0, 0, 7, 2, 0, 13, 7, 2, 0, 7, 11];
    let vec2 = vec![0, 0, 0, 2, 0, 13, 0, 2, 0, 0, 11];
    let row = make_row(1, vec1, |_| false);
    assert_eq!(row, Row {
      index: 1,
      default: 7,
      num_default: 7,
      values: vec2,
    });
  }

  #[test]
  fn write_rows() {
    let mut state = CompressState::default();
    let row1 = make_row(1, vec![0, 0, 0, 2, 3, 2, 0], |_| false);

    assert_eq!(
      -4,
      write_row(&mut state, &row1, 10, false, true)
    );

    assert_eq!(
      vec![3, 0, 0],
      state.parse_table,
    );

    assert_eq!(
      vec![14, -1, -1],
      state.check_table,
    );

    let row2 = make_row(1, vec![5, 8, 5, 0, -3, 0, 2, -1], |_| false);

    assert_eq!(
      0,
      write_row(&mut state, &row2, 10, false, true)
    );

    assert_eq!(
      vec![3, 8, 0, 0, -3, 0, 2, -1],
      state.parse_table,
    );

    assert_eq!(
      vec![14, 11, -1, -1, 14, -1, 16, 17],
      state.check_table,
    );

    let row3 = make_row(1, vec![0; 8], |_| false);

    assert_eq!(
      -8,
      write_row(&mut state, &row3, 10, false, true)
    );

    assert_eq!(
      vec![3, 8, 0, 0, -3, 0, 2, -1],
      state.parse_table,
    );

    assert_eq!(
      vec![14, 11, -1, -1, 14, -1, 16, 17],
      state.check_table,
    );

    let row4 = make_row(1, vec![3, 1, 1, 6, 7, 1, 1, 55], |_| false);

    assert_eq!(
      5,
      write_row(&mut state, &row4, 10, false, true)
    );

    assert_eq!(
      vec![3, 8, 0, 0, -3, 3, 2, -1, 6, 7, 0, 0, 55],
      state.parse_table,
    );

    assert_eq!(
      vec![14, 11, -1, -1, 14, 10, 16, 17, 13, 14, -1, -1, 17],
      state.check_table,
    );

    let row5 = make_row(1, vec![0, 0, 12, 6, 5, 5, 0, 5], |_| false);

    assert_eq!(
      0,
      write_row(&mut state, &row5, 10, false, true)
    );

    assert_eq!(
      vec![3, 8, 12, 6, -3, 3, 2, -1, 6, 7, 0, 0, 55],
      state.parse_table,
    );

    assert_eq!(
      vec![14, 11, 12, 13, 14, 10, 16, 17, 13, 14, -1, -1, 17],
      state.check_table,
    );

    let row6 = make_row(1, vec![1, 2, 3, 0, 4, 4, 0, 0], |_| false);

    assert_eq!(
      13,
      write_row(&mut state, &row6, 10, false, true)
    );

    assert_eq!(
      vec![3, 8, 12, 6, -3, 3, 2, -1, 6, 7, 0, 0, 55, 1, 2, 3, 0, 0, 0, 0, 0],
      state.parse_table,
    );

    assert_eq!(
      vec![14, 11, 12, 13, 14, 10, 16, 17, 13, 14, -1, -1, 17, 10, 11, 12, -1, -1, -1, -1, -1],
      state.check_table,
    );
  }

  #[test]
  fn action_table() {
    let mut state = CompressState::default();

    let (disp, default) = compress_action_table(
      &mut state,
      vec![
        vec![0, 1, -4, -4, 2],
        vec![3, 4, 0, 5, 0],
        vec![6, 0, 7, 0, -12],
        vec![0, 0, 0, 8, 0],
        vec![9, -3, 0, 0, 10],
      ],
    );

    assert_eq!(
      state.parse_table,
      vec![3, 4, 1, 5, 6, 2, 7, 9, 8, 0, 0, 10],
    );

    assert_eq!(
      state.check_table,
      vec![0, 1, 1, 3, 0, 4, 2, 0, 3, -1, -1, 4],
    );

    assert_eq!(
      disp,
      vec![1, 0, 4, 5, 7],
    );

    assert_eq!(
      default,
      vec![-4, 0, -12, 0, -3],
    );
  }

  #[test]
  fn goto_table() {
    let mut state = CompressState::default();

    let (disp, default) = compress_goto_table(
      &mut state,
      vec![
        vec![0, 1, 4, 4, 2],
        vec![3, 4, 9, 5, 9],
        vec![6, 12, 7, 0, 12],
        vec![0, 1, 1, 8, 0],
        vec![9, 3, 0, 3, 10],
      ],
      100
    );

    assert_eq!(
      state.parse_table,
      vec![3, 4, 1, 5, 6, 2, 7, 9, 8, 0, 0, 10],
    );

    assert_eq!(
      state.check_table,
      vec![101, 101, 100, 101, 102, 100, 102, 104, 103, -1, -1, 104],
    );

    assert_eq!(
      disp,
      vec![1, 0, 4, 5, 7],
    );

    assert_eq!(
      default,
      vec![4, 9, 12, 1, 3],
    );
  }

  #[test]
  fn action_table_duplicate_rows() {
    let mut state = CompressState::default();

    let (disp, default) = compress_action_table(
      &mut state,
      vec![
        vec![0, 1, -4, -4, 2],
        vec![3, 4, 0, 5, 0],
        vec![6, 0, 7, 0, -12],
        vec![0, 0, 0, 8, 0],
        vec![9, -3, 0, 0, 10],
        vec![9, 0, 0, -1, 10],
        vec![9, 0, 0, 0, 0],
      ],
    );

    assert_eq!(
      state.parse_table,
      vec![3, 4, 1, 5, 6, 2, 7, 9, 8, 9, 0, 10, 0, 0],
    );

    assert_eq!(
      state.check_table,
      vec![0, 1, 1, 3, 0, 4, 2, 0, 3, 0, -1, 4, -1, -1],
    );

    assert_eq!(
      disp,
      vec![1, 0, 4, 5, 7, 7, 9],
    );

    assert_eq!(
      default,
      vec![-4, 0, -12, 0, -3, -1, 0],
    );
  }
}