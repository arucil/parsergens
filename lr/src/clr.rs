use crate::builder::{LrComputation, StateStore, KernelItemSet, State, Item};

pub struct ClrComputation;

impl LrComputation for ClrComputation {
  type StateKey = Vec<Item>;

  /// `kernel_item_set` is sorted by the field `key`.
  ///
  /// Returns whether the state has changed.
  fn store_state(
    states: &mut StateStore<Self::StateKey>,
    kernel_item_set: KernelItemSet,
  ) -> (u32, bool) {
    if let Some(i) = states.get_index_of(&kernel_item_set) {
      (i as u32, false)
    } else {
      let state_ix = states.insert_full(
        kernel_item_set.clone(),
        State::new(kernel_item_set),
      ).0 as u32;

      (state_ix, true)
    }
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::builder::*;
  use insta::{assert_snapshot, assert_debug_snapshot};
  use grammar::LoweredGrammar;
  use crate::augment;

  fn prepare(input: &str) -> LoweredGrammar {
    let grammar = grammar::build(input).unwrap();
    let grammar = grammar.lower();
    let grammar = augment::augment(grammar);

    grammar
  }

  fn merge_action_goto((action, goto): (Vec<Vec<i32>>, Vec<Vec<u32>>)) -> Vec<Vec<i32>> {
    action.into_iter().enumerate().map(|(state, mut row1)| {
      row1.extend((0..goto.len()).map(|nt| goto[nt][state] as i32));
      row1
    })
    .collect()
  }

  static SIMPLE: &str = r#"
%token c "c"
%token d "d"
%start S
S = C C
C = c C
  | d
  "#;

  #[test]
  fn simple_states() {
    let grammar = prepare(SIMPLE);
    let mut builder = Builder::<ClrComputation>::new(&grammar);
    let start_nts = gen_states(&mut builder);

    assert_snapshot!(builder.states(&start_nts));
  }

  #[test]
  fn simple_action_goto() {
    let grammar = prepare(SIMPLE);
    let mut builder = Builder::<ClrComputation>::new(&grammar);
    gen_states(&mut builder);
    let tables = merge_action_goto(gen_tables(&builder).unwrap());

    assert_debug_snapshot!(tables);
  }

  static EPSILON: &str = r#"
%token plus "+"
%token mult "*"
%token num "1"
%token lparen "("
%token rparen ")"

%start E

E = T E'
E' = plus T E'
   | ()
T = F T'
T' = mult F T'
   | ()
F = num
  | lparen E rparen
  "#;

  #[test]
  fn epsilon_states() {
    let grammar = prepare(EPSILON);
    let mut builder = Builder::<ClrComputation>::new(&grammar);
    let start_nts = gen_states(&mut builder);

    assert_snapshot!(builder.states(&start_nts));
  }

  #[test]
  fn epsilon_action_goto() {
    let grammar = prepare(EPSILON);
    let mut builder = Builder::<ClrComputation>::new(&grammar);
    let _start_nts = gen_states(&mut builder);
    let tables = gen_tables(&builder).unwrap();
    let tables = merge_action_goto(tables);

    assert_debug_snapshot!(tables);
  }
}