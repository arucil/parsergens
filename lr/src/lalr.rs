use std::collections::HashMap;
use grammar::{ LoweredGrammar, TokenId, BiMap };
use bit_set::BitSet;
use crate::Error;
use crate::ffn::Ffn;
use crate::builder::LrCalculation;
use crate::clr::{ClrCalc, Lr1Item};

pub enum LalrCalc {}

impl LrCalculation for LalrCalc {
  type Item = Lr1Item;

  fn start_item(
    start_prod_ix: usize,
    eof_token: TokenId,
  ) -> Lr1Item {
    ClrCalc::start_item(start_prod_ix, eof_token)
  }

  fn next_item(
    item: &Lr1Item
  ) -> Lr1Item {
    ClrCalc::next_item(item)
  }

  #[inline(always)]
  fn closure_step<F>(
    grammar: &LoweredGrammar,
    ffn: &Ffn,
    prev: &Lr1Item,
    action: F
  )
    where F: FnMut(Lr1Item)
  {
    ClrCalc::closure_step(grammar, ffn, prev, action)
  }

  #[inline(always)]
  fn reduce_tokens<F>(
    grammar: &LoweredGrammar,
    ffn: &Ffn,
    item: &Lr1Item,
    action: F,
  ) -> Result<(), Error>
    where F: FnMut(u32) -> Result<(), Error>
  {
    ClrCalc::reduce_tokens(grammar, ffn, item, action)
  }

  fn merge_states(
    _grammar: &LoweredGrammar,
    states: &mut BiMap<BitSet, u32>,
    items: &mut BiMap<Lr1Item, usize>,
  ) -> Option<HashMap<u32, u32>> {
    let mut new_states = BiMap::<BitSet, u32>::new();
    let mut new_items = BiMap::new();
    let mut map = HashMap::new();

    for (state, state_id) in &*states {
      let new_state = state.iter().map(|item| {
        let item = items.get_by_right(&item).unwrap();
        store_item(&mut new_items, item)
      }).collect::<BitSet>();
      let new_state_id = if let Some(id) = new_states.get_by_left(&new_state) {
        *id
      } else {
        let id = new_states.len() as u32;
        new_states.insert(new_state, id);
        id
      };
      map.insert(*state_id, new_state_id);
    }

    *items = new_items;
    *states = new_states;
    Some(map)
  }
}

fn store_item(
  items: &mut BiMap<Lr1Item, usize>,
  item: &Lr1Item
) -> usize {
  let item = Lr1Item {
    token: std::u32::MAX,
    ..*item
  };
  if let Some(&item) = items.get_by_left(&item) {
    item
  } else {
    let len = items.len();
    items.insert(item, len);
    len
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use insta::{assert_debug_snapshot, assert_snapshot};
  use grammar::{ TokenId, LoweredGrammar };
  use crate::ffn::Ffn;
  use crate::ffn;
  use crate::augment;
  use crate::builder::Builder;

  fn prepare(input: &str) -> (LoweredGrammar, TokenId, Ffn) {
    let grammar = grammar::build(input).unwrap();
    let grammar = grammar.lower();
    let (grammar, eof_token) = augment::augment(grammar);
    let ffn = ffn::compute(&grammar);

    (grammar, eof_token, ffn)
  }

  static SIMPLE: &str = r#"
%token c "c"
%token d "d"

%start S

S = C C
C = c C
  | d
  "#;

  #[test]
  fn simple_states() {
    let (grammar, eof_token, ffn) = prepare(SIMPLE);
    let mut builder = Builder::<LalrCalc>::new(&grammar, eof_token, ffn);

    builder.build().unwrap();

    assert_snapshot!(builder.states());
  }

  #[test]
  fn simple_action_goto() {
    let (grammar, eof_token, ffn) = prepare(SIMPLE);
    let mut builder = Builder::<LalrCalc>::new(&grammar, eof_token, ffn);

    builder.build().unwrap();

    assert_debug_snapshot!(builder.action_goto());
  }

  static TIGER: &str = r##"
%token EQ          "="
%token NEQ         "<>"
%token LT          "<"
%token GT          ">"
%token LE          "<="
%token GE          ">="
%token ASSIGN      ":="
%token COLON       ":"
%token COMMA       ","
%token DOT         "."
%token SEMI        ";"
%token PLUS        "+"
%token MINUS       "-"
%token TIMES       "*"
%token DIV         "/"
%token MOD         "%"
%token AND         "&"
%token OR          "|"
%token LBRACE      "{"
%token RBRACE      "}"
%token LPAREN      "("
%token RPAREN      ")"
%token LBRACK      "["
%token RBRACK      "]"
%token TYPE        "type"
%token INT         "int"
%token STRING      "string"
%token ARRAY       "array"
%token OF          "of"
%token VAR         "var"
%token NIL         "nil"
%token FUNCTION    "function"
%token IF          "if"
%token THEN        "then"
%token ELSE        "else"
%token FOR         "for"
%token TO          "to"
%token DO          "do"
%token WHILE       "while"
%token BREAK       "break"
%token LET         "let"
%token IN          "in"
%token END         "end"
%token IDENT       /[a-zA-Z][\w_]*/
%token LIT_INT     /\d+/
%token LIT_STR     /"([^\\\n]|\\([nt"\\]|\d\d\d))*"/

%skip /[ \n]+/

%non-assoc    ELSE
%right-assoc  NEG
%left-assoc   MULT
%left-assoc   ADD
%non-assoc    REL
%non-assoc    EQ
%right-assoc  AND
%right-assoc  OR
%non-assoc    ASSIGN
%non-assoc    ARRAY
%non-assoc    CONTROL

%start program

program = expr

decls = (decl)*

decl = type-decl | var-decl | fun-decl

type-decl = TYPE type-id EQ type

type =
    IDENT
  | LBRACE type-fields RBRACE

type-fields =
    ()
  | IDENT COLON type-id (COMMA IDENT COLON type-id)*

type-id = predef-type | IDENT

predef-type = STRING | INT

var-decl = VAR IDENT (COLON type-id)? ASSIGN expr

fun-decl = FUNCTION IDENT LPAREN type-fields RPAREN (COLON type-id)? EQ expr

lvalue = IDENT lvalue-suffix

lvalue-suffix =
    ()
  | DOT IDENT lvalue-suffix
  | LBRACK expr RBRACK lvalue-suffix

expr =
    lvalue
  | lvalue ASSIGN expr %prec ASSIGN
  | NIL
  | LPAREN seq-exprs RPAREN
  | LIT_INT
  | LIT_STR
  | MINUS expr %prec NEG
  | IDENT LPAREN arg-exprs RPAREN
  | expr MOD expr %prec MULT
  | expr DIV expr %prec MULT
  | expr TIMES expr %prec MULT
  | expr PLUS expr %prec ADD
  | expr MINUS expr %prec ADD
  | expr GT expr %prec REL
  | expr LT expr %prec REL
  | expr GE expr %prec REL
  | expr LE expr %prec REL
  | expr EQ expr %prec EQ
  | expr NEQ expr %prec EQ
  | expr AND expr %prec AND
  | expr OR expr %prec OR
  | type-id LBRACE field-init-exprs RBRACE
  | IDENT LBRACK expr RBRACK OF expr %prec ARRAY
  | predef-type LBRACK expr RBRACK OF expr %prec ARRAY
  | IF expr THEN expr ELSE expr %prec ELSE
  | IF expr THEN expr %prec CONTROL
  | WHILE expr DO expr %prec CONTROL
  | FOR IDENT ASSIGN expr TO expr DO expr %prec CONTROL
  | BREAK
  | LET decls IN seq-exprs END

seq-exprs =
    ()
  | expr (SEMI expr)*

arg-exprs =
    ()
  | expr (COMMA expr)*

field-init-exprs =
    ()
  | IDENT EQ expr (COMMA IDENT EQ expr)*
  "##;

  #[test]
  #[ignore]
  fn tiger_states() {
    let (grammar, eof_token, ffn) = prepare(TIGER);
    let mut builder = Builder::<LalrCalc>::new(&grammar, eof_token, ffn);

    builder.build().unwrap();

  }
}