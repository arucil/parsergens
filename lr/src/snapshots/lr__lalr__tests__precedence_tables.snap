---
source: lr/src/lalr.rs
expression: tables
---
CompressedTables {
    parse_table: [
        5,
        6,
        7,
        8,
        7,
        15,
        5,
        6,
        7,
        8,
        2,
        5,
        6,
        7,
        3,
        0,
        4,
        6,
        7,
        1,
        0,
        9,
        0,
        0,
        11,
        12,
        13,
        14,
    ],
    action_disp: [
        10,
        6,
        10,
        10,
        -8,
        10,
        10,
        10,
        10,
        -8,
        0,
        16,
        2,
        2,
        11,
        -8,
    ],
    goto_disp: [
        19,
        -16,
    ],
    action_default: [
        0,
        -8,
        0,
        0,
        -7,
        0,
        0,
        0,
        0,
        -5,
        0,
        -1,
        -2,
        -3,
        -4,
        -6,
    ],
    goto_default: [
        10,
        0,
    ],
    check_table: [
        0,
        1,
        2,
        3,
        2,
        5,
        0,
        1,
        2,
        3,
        0,
        0,
        1,
        2,
        4,
        -1,
        6,
        1,
        2,
        8,
        -1,
        8,
        -1,
        -1,
        8,
        8,
        8,
        8,
    ],
}
